android/app/src/main/java/com/epfl/dedis/hbt/data/UserDataSource.kt

```diff
@@ -1,25 +1,51 @@
 package com.epfl.dedis.hbt.data
 
+import android.content.SharedPreferences
 import com.epfl.dedis.hbt.data.model.Role
 import com.epfl.dedis.hbt.data.model.User
 import com.epfl.dedis.hbt.data.model.Wallet
+import com.epfl.dedis.hbt.utility.json.JsonService
+import com.epfl.dedis.hbt.utility.json.JsonType.USER_DATA
 import javax.inject.Inject
 import javax.inject.Singleton
 
 /**
  * Class that handles authentication w/ login credentials and retrieves user information.
  */
 @Singleton
-class UserDataSource @Inject constructor() {
+class UserDataSource @Inject constructor(
+    private val sharedPref: SharedPreferences,
+    private val jsonService: JsonService
+) {
+
+    private val usernamesKey: String = "users"
 
     private val users: MutableMap<String, User> = mutableMapOf()
     private val wallets: MutableMap<User, Wallet> = mutableMapOf()
 
-    // TODO: add a datastore for persistence,
-    //  as described in https://developer.android.com/topic/libraries/architecture/datastore
+    init {
+        // Load data from preferences
+        val usernames = sharedPref.getStringSet(usernamesKey, setOf())!!
+        usernames.forEach {
+            val user = getUserData(it, sharedPref, jsonService)
+            users[it] = user
+            wallets[user] = Wallet.newInstance()
```

Comment:
if a user exists, the wallet should already exist too, so we should load the existing wallet instead, shouldn't we ?